File: chat_hub/core/services/chat/dialogue_service.go
package services

import (
	entity "chat_hub/core/domain/entities"
	"chat_hub/core/domain/enums"
	"chat_hub/infrastructure/repository"
	"database/sql"
	"fmt"
	"log"
	"strconv"
	"time"

	"github.com/google/uuid"
)

type DialogueService struct {
	db *sql.DB

	dialogueRepo *repository.DialogueRepository
}

func NewDialogueService(db *sql.DB) *DialogueService {
	return &DialogueService{
		db:           db,
		dialogueRepo: repository.NewDialogueRepository(db),
	}
}

func (s *DialogueService) GetOrCreateDialogue(userId, dialogueId, msgType string) (entity.UserDialogueEntity, error) {
	if dialogueId != "" {
		dialogue, err := s.GetDialogueById(userId, dialogueId)
		if err != nil {
			return entity.UserDialogueEntity{}, fmt.Errorf("failed to retrieve dialogue: %w", err)
		}
		return dialogue, nil
	}

	dialogue, err := s.CreateDialogueIfNotExists(userId, msgType)
	if err != nil {
		return entity.UserDialogueEntity{}, fmt.Errorf("failed to create dialogue: %w", err)
	}
	return dialogue, nil
}

func (s *DialogueService) CreateDialogueIfNotExists(userId, dialogueType string) (entity.UserDialogueEntity, error) {
	log.Println("Checking if dialogue exists for user:", userId, "with dialog type:", dialogueType)

	dialogue, err := s.GetDialogueByUserIdAndType(userId, dialogueType)
	if err != nil && err != sql.ErrNoRows {
		log.Println("Error retrieving dialogue from repository:", err)
		return entity.UserDialogueEntity{}, err
	}

	if err == nil && dialogue.ID != "" {
		log.Println("Dialogue already exists for user:", userId, "with dialog type:", dialogueType)
		return dialogue, nil
	}

	return s.CreateDialogue(userId, dialogueType)
}

func (s *DialogueService) CreateDialogue(userId, dialogueType string) (entity.UserDialogueEntity, error) {
	userIdF, err := strconv.ParseInt(userId, 10, 64)
	if err != nil {
		log.Println("Error parsing user ID:", err)
		return entity.UserDialogueEntity{}, err
	}

	dialogue := entity.UserDialogueEntity{
		ID:             uuid.New().String(),
		UserID:         userIdF,
		DialogueName:   enums.GaiaIntroductionDialogue,
		DialogueType:   dialogueType,
		DialogueStatus: enums.ACTIVE,
		Metadata:       "{}",
		CreatedAt:      time.Now(),
		UpdatedAt:      time.Now(),
	}

	createdDialogue, err := s.dialogueRepo.CreateDialogue(dialogue)
	if err != nil {
		log.Println("Error creating dialogue in repository:", err)
		return entity.UserDialogueEntity{}, err
	}
	log.Println("Storing dialogue in repository:", dialogue)

	return createdDialogue, nil
}

func (s *DialogueService) GetDialogueByUserIdAndType(userId, dialogueType string) (entity.UserDialogueEntity, error) {
	log.Println("Retrieving dialogue for user:", userId, "with dialog type:", dialogueType)

	dialogue, err := s.dialogueRepo.GetDialogueByUserIdAndType(userId, dialogueType)
	if err != nil && err != sql.ErrNoRows {
		log.Println("Error retrieving dialogue from repository:", err)
		return entity.UserDialogueEntity{}, err
	}

	return dialogue, err
}

func (s *DialogueService) GetDialogueById(userId, dialogueId string) (entity.UserDialogueEntity, error) {
	log.Println("Retrieving dialogue by ID:", dialogueId)

	dialogue, err := s.dialogueRepo.GetDialogueById(userId, dialogueId)
	if err != nil && err != sql.ErrNoRows {
		log.Println("Error retrieving dialogue from repository:", err)
		return entity.UserDialogueEntity{}, err
	}

	return dialogue, err 
}

File: ai_core/core/service/integration/dialogue_service.py
import uuid
import time

from core.domain.enums.enum import DialogueEnum, ActiveEnum
from core.domain.entities.user_dialogue import UserDialogue
from infrastructure.repository.dialogue_repository import user_dialogue_repository 


class DialogueService:
    """Service layer for user dialogues, interfacing with the UserDialogueRepository."""
    
    async def get_or_create_dialogue(self, userId: str, dialogue_id: str, msg_type: str) -> UserDialogue:
        try:
            if dialogue_id:
                return await self._get_dialogue_by_id(dialogue_id)
            return await self._create_dialogue_if_not_exists(userId, msg_type)
        except Exception as e:
            print(f"Error in get_or_create_dialogue: {e}")
            return None

    async def _get_dialogue_by_id(self, user_id: str, dialogue_id: str) -> UserDialogue:
        return await user_dialogue_repository.get_dialogue_by_id(user_id, dialogue_id)

    async def _create_dialogue_if_not_exists(self, user_id: str, dialogue_type: str) -> UserDialogue:
        dialogue = await self._get_dialogue_by_user_id_and_type(user_id, dialogue_type)
        if dialogue is not None and dialogue.id is not None:
            return dialogue
        return self._create_dialogue(user_id, dialogue_type)

    async def _get_dialogue_by_user_id_and_type(self, user_id: str, dialogue_type: str) -> UserDialogue:
        return await user_dialogue_repository.get_dialogue_by_user_id_and_type(user_id, dialogue_type)
    
    async def _create_dialogue(self, user_id: str, dialogue_type: str) -> UserDialogue:
        new_dialogue = UserDialogue(
            id=uuid.uuid4(),
            user_id=int(user_id),
            dialogue_name=DialogueEnum.GAIA_INTRODUCTION.value,
            dialogue_type=dialogue_type,
            dialogue_status=ActiveEnum.ACTIVE.value,
            metadata={},
            created_at=time.time(),
            updated_at=time.time(),
        )
        return await user_dialogue_repository.create_dialogue(new_dialogue)
    
    async def get_all_dialogues_by_user_id(self, user_id: str, size: int, cursor: str | None) -> tuple[list[UserDialogue], bool]:
        try:
            dialogues, has_more = await user_dialogue_repository.get_all_dialogues_by_user_id(user_id, size, cursor)
            if len(dialogues) == 0:
                print("No dialogues found for user_id:", user_id)
                return [], False
            return dialogues, has_more
        except Exception as e:
            raise e

dialogue_service = DialogueService()
